backend/app/models/sample_type.rb

```diff
@@ -0,0 +1,284 @@
+    sql = "
+      select * from sample_types order by name
+    "
+    SampleType.find_by_sql sql
+  end
+
+  def self.find_id(id)
+    sql = "
+      select * from sample_types where id = #{id} limit 1
+    "
+    (SampleType.find_by_sql sql)[0]
+  end
+
+  # Return details for a specific sample types.
+  #
+  # @param id [Int] the id of the sample type
+  #
+  # @return the sample types
+  def self.details(id)
+    # GET FEILD TYPES
+    sql = "
+      select *, null as 'allowable_field_types'
+      from field_types ft
+      where ft.parent_class = 'SampleType' and ft.parent_id = #{id}
+      order by ft.name
+    "
+    field_types = FieldType.find_by_sql sql
+
+    # GET SAMPLE_OPTIONS FOR FIELD_TYPE == "SAMPLE"
+    # MAKES <N> CALLS TO THE DB BUT <N> SHOULD NOT BE LARGE
+    field_types.each do |ft|
+      if ft.ftype == "sample"
+        sql = "
+          select aft.id, aft.field_type_id, aft.sample_type_id, st.name
+          from allowable_field_types aft
+          inner join sample_types st on st.id = aft.sample_type_id
+          where aft.field_type_id = #{ft.id}
+          order by st.name
+        "
+        allowable_field_types = AllowableFieldType.find_by_sql sql
+
+        ft = ft.update( {allowable_field_types: allowable_field_types} )
+      end
+    end
+
+    # GET INVENTORY
+    # TODO -- IMPLEMENT THIS WHEN IMPLEMENT INVENTORY
+    inventory = 0
+
+    # GET OBJECT TYPES (CONTAINERS)
+    sql = "
+      select * from object_types where sample_type_id = #{id} order by name
+    "
+    object_types = ObjectType.find_by_sql sql
+
+    return { field_types: field_types, inventory: inventory, object_types: object_types }
+  end
+
+  # Create a sample types.
+  #
+  # @param st [Hash] the sample type
+  #
+  # @option st[:name] [String] the name of the sample type
+  # @option st[:description] [String] the description of the sample type
+  # @option st[:field_types] [Hash] the field_type attributes associated with the sample type
+  #
+  # feild_types = {
+  #   name [String] name
+  #   ftype [String]  ftype
+  #   required [String] required - interpreted as Boolen
+  #   array [String] array - interpreted as Boolean
+  #   choices [String] choices
+  #   allowable_field_types [Array] array of allowable field types (for ftype = "sample")
+  # }
+  #
+  # allowable_field_type = {
+  #   sample_type_id [Int] id of the allowable field type
+  # }
+  #
+  # @return the sample type
+  def self.create(st)
+    name = Input.text(st[:name])
+    description = Input.text(st[:description])
+
+    # CREATE TEH SAMPLE TYPE
+    sample_type_new = SampleType.new
+    sample_type_new.name = name
+    sample_type_new.description = description
+
+    valid = sample_type_new.valid?
+    return false, sample_type_new.errors if !valid
+
+    # SAVE THE SAMPLE TYPE IF IT IS VALID
+    sample_type_new.save
+
+    # SAVE FIELD TYPE IF THERE IS A FIELD NAME
+    if st[:field_types].kind_of?(Array)
+      st[:field_types].each do |ft|
```

Comment:
Why is this deserialization code in `SampleType`?
It is creating a `FieldType`, so it belongs there instead

Reply:
Because when you create a sample type you send a complex form that ALSO creates field types <and> allowable_field_types. It is a complex object. This seems clean to me (it does what you ask), but If you have ideas on how to change it we can discuss in our 1-1.

Reply:
again, we discussed this, but just to document the point.

This code is distinct from both the frontend and the database. If you have a serialized FieldType from which a FieldType object (and corresponding record) needs to be created, it is cleaner to have the code for deserializing in the FieldType class.  This is just responsibility allocation in OO.